### PR TITLE
Endpoints GRPC Python Sample

### PR DESCRIPTION
**Pull request by ryanmats**
Note: There are linting issues withe helloworld_pb2.py and helloworld_pb2_grpc.py but we plan on ignoring these issues since they are generated code.

### CODE REVIEW

File: endpoints/getting-started-grpc/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+WORKDIR /hello
+
+RUN pip install -r requirements.txt
+
+EXPOSE 8000
+
+# Copy Server
+COPY . /hello/
+
```
Comment (jeffmendoza) [inline endpoints/getting-started-grpc/Dockerfile]: In general, dont use `ADD`. Should look something like:

```
COPY requirements.txt /tmp/
RUN pip install --requirement /tmp/requirements.txt
COPY . /tmp/
```
see: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

File: endpoints/getting-started-grpc/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ENV VIRTUAL_ENV -p python3.5 /env
+ENV PATH /env/bin:$PATH
+
+ADD . /hello/
+
+WORKDIR /hello
+
+RUN pip install -r requirements.txt
+
+EXPOSE 8000
```
Comment (jeffmendoza) [inline endpoints/getting-started-grpc/Dockerfile]: I believe this container is exposing 50051, but listing it is not necessary

File: endpoints/getting-started-grpc/helloworld/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+[This code's documentation lives on the grpc.io site.](http://www.grpc.io/docs/quickstart/python.html)
```
Comment (jeffmendoza) [inline endpoints/getting-started-grpc/helloworld/README.md]: Why keep the code under `helloworld/`? Doesn't make sense with only one sample. Either way this README is not needed.

File: endpoints/getting-started-grpc/protos/BUILD [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2017, Google Inc.
```
Comment (jeffmendoza) [inline endpoints/getting-started-grpc/protos/BUILD]: I don't think Python devs use bazel too much. Makes sense to remove this.

File: endpoints/getting-started-grpc/protos/helloworld.proto [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "io.grpc.examples.helloworld";
+option java_outer_classname = "HelloWorldProto";
+option objc_class_prefix = "HLW";
+
```
Comment (jeffmendoza) [inline endpoints/getting-started-grpc/protos/helloworld.proto]: You can remove these options for java and objc

File: endpoints/getting-started-grpc/Dockerfile (hunk 2 above)
Comment (ryanmats) [inline endpoints/getting-started-grpc/Dockerfile]: done

File: endpoints/getting-started-grpc/Dockerfile (hunk 1 above)
Comment (ryanmats) [inline endpoints/getting-started-grpc/Dockerfile]: done

Comment (ryanmats) [inline endpoints/getting-started-grpc/Dockerfile]: done

File: endpoints/getting-started-grpc/protos/BUILD (hunk 4 above)
Comment (ryanmats) [inline endpoints/getting-started-grpc/protos/BUILD]: done

File: endpoints/getting-started-grpc/protos/helloworld.proto (hunk 5 above)
Comment (ryanmats) [inline endpoints/getting-started-grpc/protos/helloworld.proto]: done

File: endpoints/getting-started-grpc/helloworld/README.md (hunk 3 above)
Comment (ryanmats) [inline endpoints/getting-started-grpc/helloworld/README.md]: removed the helloworld/ subdirectory and moved files to getting-started-grpc/ directory

File: endpoints/getting-started-grpc/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# Create a virtualenv for dependencies. This isolates these packages from
+# system-level packages.
+RUN virtualenv /env
+
+# Setting these environment variables are the same as running
+# source /env/bin/activate.
+ENV VIRTUAL_ENV -p python3.5 /env
+ENV PATH /env/bin:$PATH
+
+COPY requirements.txt /tmp/
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/Dockerfile]: done

File: endpoints/getting-started-grpc/greeter_client.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""The Python implementation of the GRPC helloworld.Greeter client."""
+
+from __future__ import print_function
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/greeter_client.py]: done

File: endpoints/getting-started-grpc/greeter_server.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        return helloworld_pb2.HelloReply(
+            message='Hello again, %s!' % request.name)
+
+
+def serve():
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+    server.add_insecure_port('[::]:50051')
+    server.start()
+    try:
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/greeter_server.py]: done

File: endpoints/getting-started-grpc/greeter_server.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        while True:
+            time.sleep(_ONE_DAY_IN_SECONDS)
+    except KeyboardInterrupt:
+        server.stop(0)
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/greeter_server.py]: http://www.grpc.io/grpc/python/grpc.html#grpc.Server.stop

"grace – A duration of time in seconds or None. If a duration of time in seconds, the time to allow existing RPCs to complete before being aborted by this Server’s stopping. If None, all RPCs will be aborted immediately and this method will block until this Server is completely stopped."

Let me know what you want to do with this

File: endpoints/getting-started-grpc/greeter_client.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+def run(host, api_key):
+    channel = grpc.insecure_channel(host)
+    stub = helloworld_pb2_grpc.GreeterStub(channel)
+    metadata = []
+    if api_key:
+        metadata.append(('x-api-key', api_key))
+    response = stub.SayHello(
+        helloworld_pb2.HelloRequest(name='you'), metadata=metadata)
+    print("Greeter client received: " + response.message)
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/greeter_client.py]: done

File: endpoints/getting-started-grpc/greeter_server.py (hunk 9 above)
Comment (jonparrott) [inline endpoints/getting-started-grpc/greeter_server.py]: use `stop(grace=0)`.

Comment (ryanmats) [inline endpoints/getting-started-grpc/greeter_server.py]: done

File: endpoints/getting-started-grpc/README.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+   just wish to regenerate these files, run:
+
+    ```bash
+    python -m grpc_tools.protoc -I protos --python_out=. --grpc_python_out=. protos/helloworld.proto
+    ```
+
+1. Generate the `out.pb` from the proto file:
+
+    ```bash
+    python -m grpc_tools.protoc --include_imports --include_source_info -I protos protos/helloworld.proto --descriptor_set_out out.pb
```
Comment (ryanmats) [inline endpoints/getting-started-grpc/README.md]: @jonparrott out.pb is what is used in the Java canonical sample (https://github.com/GoogleCloudPlatform/java-docs-samples/tree/master/endpoints/getting-started-grpc0). do you wanna change it to something else?

Comment (jonparrott) [inline endpoints/getting-started-grpc/README.md]: It's fine, I was just questioning Jeff, but he's on leave now so it's whatever. :)

**Comment by jonparrott** [comment]
@dpebot can you merge when travis passes?

**Comment by jonparrott** [comment]
@ryanmats nox should pass lint now. :)

**Comment by jonparrott** [comment]
@dpebot will you merge when travis is green? pretty please?

**Pull request merged**